pennylane/devices/qutrit_mixed/measure.py

```diff
@@ -0,0 +1,309 @@
+    measurementprocess: StateMeasurement, state: TensorLike, is_state_batched: bool = False
+) -> TensorLike:
+    """Get the state or reduced density matrix.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (TensorLike): state to apply the measurement to
+        is_state_batched (bool): whether the state is batched or not
+
+    Returns:
+        TensorLike: the resulting resquared density_matrix
+    """
+    wires = measurementprocess.wires
+    if not wires:
+        return resquare_state(state, get_num_wires(state, is_state_batched))
+
+    num_state_wires = get_num_wires(state, is_state_batched)
+    wires_to_trace = [x + is_state_batched for x in range(num_state_wires) if x not in wires]
+
+    for i, wire_to_trace in enumerate(wires_to_trace):
+        axis1 = wire_to_trace - i
+        axis2 = axis1 + num_state_wires - i
+
+        print(axis1, axis2)
+
+        state = math.trace(state, axis1=axis1, axis2=axis2)
+
+    return resquare_state(state, len(wires))
+
+
+def _get_probs(state, num_wires, is_state_batched: bool = False):
+    """Finds the probabilities of measuring states in the computational basis
+
+    Args:
+        state (array[complex]): Input quantum state
+        num_wires (int): The number of wires the state represents
+        is_state_batched (bool): Boolean representing whether the state is batched or not
+
+    Returns:
+        Tensorlike: Vector representing probability of each state
+    """
+    # get the final shape
+    final_shape = (
+        (state.shape[0], QUDIT_DIM**num_wires) if is_state_batched else (QUDIT_DIM**num_wires,)
+    )
+
+    # probs are diagonal elements
+    # using einsum since diagonal function axis selection parameter names
+    # are not consistent across interfaces
+    indices = alphabet[:num_wires]
+    probs = math.einsum(f"...{indices * 2}->...{indices}", state)
+    # take the real part so probabilities are not shown as complex numbers
+    # also reshape to flat
+    probs = math.reshape(math.real(probs), final_shape)
+    return math.where(probs < 0, -probs, probs)
+
+
+def calculate_probability(
+    measurementprocess: StateMeasurement, state: TensorLike, is_state_batched: bool = False
+) -> TensorLike:
+    """Find the probability of measuring states.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (TensorLike): state to apply the measurement to
+        is_state_batched (bool): whether the state is batched or not
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    for op in measurementprocess.diagonalizing_gates():
+        state = apply_operation(op, state, is_state_batched=is_state_batched)
+
+    num_state_wires = get_num_wires(state, is_state_batched)
+    probs = _get_probs(state, num_state_wires, is_state_batched)
+
+    if mp_wires := measurementprocess.wires:
```

Comment:
What is the `:=` doing?

Reply:
it defines `mp_wires = measurementprocess.wires` and does `if measurementprocess.wires` its used elsewhere in PennyLane.

Reply:
Walruses!

Reply:
Huh! Learned something new :)